File: src/tests/ts-loss-tau.cpp
#include <cstdio>
#include <iostream>
#include <iomanip>
#include <vector>
#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>
#include <thread>
#include <algorithm>
#include <gillespie-algorithm.hpp>

// A Gillespie algorithm simulation of tumour suppressor loss using tau-leaping

using clonal_expansion::gillespie_ssa::times_to_final_vertices_tau;
using clonal_expansion::gillespie_ssa::print_kaplan_meier;

int main(int argc, char* argv[]) {
    int sample_size = 10; // default values
    int seed = 1;
    double tau = 0.1; // default tau value for tau-leaping
    if (argc < 4) {
        printf("Call this program with\n ./tsgillespie_tau seed runs tau\n");
        return 1;
    }

    // TODO multithreading can break this test!
    seed = atoi(argv[1]);
    sample_size = atoi(argv[2]);
    tau = atof(argv[3]); // set the tau value from command line argument

    // System coefficients:
    double rloh = 5e-7;
    double mu = 5e-8;
    printf("%.4f %.4f %.4f\n", rloh, mu, tau);

    clonal_expansion::Model model(5);
    model.m_migr[0][1] = mu;
    model.m_migr[0][2] = rloh;
    model.m_migr[1][3] = 0.5 * mu;
    model.m_migr[1][4] = 0.5 * rloh;
    model.m_migr[2][4] = 0.5 * mu;
    // birth and death rates:
    model.m_birth = {0, 0.05, 0.03, 0, 0};
    model.m_death = {0, 0, 0, 0, 0};
    model.m_initial_pops = {1e6, 0, 0, 0, 0};

    std::vector<int> final_vertices = {3, 4};

    // Run many simulations and store the results:
    std::vector<std::pair<double, int>> all_times;
    times_to_final_vertices_tau(model, seed, sample_size, final_vertices, tau,
                                all_times);

    std::sort(all_times.begin(), all_times.end());
    size_t study_population = all_times.size();
    double time_max = all_times.back().first;

    // Print results for both types:
    for (auto type : final_vertices) {
        std::cout << "Type " << type << ":" << std::endl;

        std::vector<double> mutant_times;
        for (const auto& result : all_times) {
            if (result.second == type) {
                mutant_times.push_back(result.first);
            }
        }

        // Guard against invalid access:
        if (mutant_times.empty()) {
            std::cout << "No results" << std::endl;
            return 2;
        }

        // Kaplan-Meier plot:
        std::cout << "age, p1, p2," << std::endl;
        print_kaplan_meier(time_max, mutant_times, sample_size);

        std::cout << std::endl;

        // Print all times:
        for (const auto& time : mutant_times) {
            printf("%.8f\n", time);
        }

        std::cout << std::endl;
    }

    return 0;
}

File: src/python/tests/ts-loss-tau.py
import pyffwd
import numpy as np
import os
import threading

def times_to_final_vertices_tau(model, seed, runs_per_thr,
                            final_vertices, tau, result_container):
    rng = pyffwd.GSL_RNG(seed)
    for _ in range(runs_per_thr):
        result = pyffwd.first_passage_time_tau(
            rng, model, final_vertices, tau)
        result_container.append(result)

def print_kaplan_meier(time_max, mutant_times, sample_size):
    pyffwd.print_kaplan_meier(time_max,
                              pyffwd.RealVector(mutant_times),
                              sample_size)

if __name__ == "__main__":
      # default values
    sample_size = 10
    seed = 1
    tau = 0.1

    import sys
    if len(sys.argv) < 4:
        print("Call this program with\n /ts_tau seed runs tau \n")
        sys.exit(1)

    # TODO multithreading can break this test
    seed = int(sys.argv[1])
    sample_size = int(sys.argv[2])
    tau = float(sys.argv[3])

    # System coefficients:
    rloh = 5e-7
    mu = 5e-8
    print("%.4f" % rloh, "%.4f" % mu, "%.4f" % tau)

    # Define the model
    model = pyffwd.Model.create_model(
        5, 
        birth_rates=[0, 0.05, 0.03, 0, 0],
        death_rates=[0, 0, 0, 0, 0],
        initial_pops=[1e6, 0, 0, 0, 0]
    )
    model.m_migr = [
        {1: mu, 2: rloh},
        {3: 0.5 * mu, 4: 0.5 * rloh},
        {4: 0.5 * mu},
        {},
        {}
    ]

    final_vertices = pyffwd.IntVector([3, 4])

    all_times = []
    times_to_final_vertices_tau(model, seed, sample_size,
                            final_vertices, tau, all_times)

    all_times.sort(key=lambda x: x[0])

    study_population = len(all_times)
    time_max = all_times[-1][0] if all_times else 0

    # Print Kaplan-Meier
    for tumour_type in final_vertices:
        print(f"Type {tumour_type}:")

        mutant_times = [result[0] for result in all_times 
                        if result[1] == tumour_type]

        if not mutant_times:
            print("No results")
            sys.exit(2)

        # Print the Kaplan-Meier
        print("age, p1, p2,", flush=True)
        print_kaplan_meier(time_max, mutant_times, sample_size)
        print()

        for time in mutant_times:
            print("%.8f" % time)

        print()
